inc/classes/subscriber/Tools/class-detect-missing-tags-subscriber.php

```diff
@@ -0,0 +1,102 @@
+<?php
+namespace WP_Rocket\Subscriber\Tools;
+
+use WP_Rocket\Event_Management\Subscriber_Interface;
+use WP_Rocket\Logger\Logger;
+
+/**
+ * Detect and report when <html>, wp_footer() and <body> tags are missing.
+ *
+ * @since  3.4.2
+ * @author Soponar Cristina
+ */
+class Detect_Missing_Tags_Subscriber implements Subscriber_Interface {
+
+	/**
+	 * @inheritDoc
+	 */
+	public static function get_subscribed_events() {
+		return [
+			'admin_notices'                      => 'notice_missing_tags',
+			'before_rocket_maybe_process_buffer' => 'maybe_missing_tags',
+		];
+	}
+
+	/**
+	 * Check if there is a missing </html> or </body> tag
+	 *
+	 * @since  3.4.2
+	 * @author Soponar Cristina
+	 *
+	 * @param string $html HTML content.
+	 */
+	public function maybe_missing_tags( $html ) {
+		Logger::info( 'START Detect_Missing_Tags_Subscriber - maybe_missing_tags ', [ 'maybe_missing_tags' ] );
+
+		// Remove all comments before testing tags. If </html> or </body> tags are commented this will identify it as a missing tag.
+		$html         = preg_replace( '/<!--([\\s\\S]*?)-->/', '', $html );
+		$missing_tags = [];
+		if ( ! preg_match( '/(<\/html>)/i', $html ) ) {
+			$missing_tags[] = esc_html__( '</html>', 'rocket' );
+			Logger::debug( 'Not found closing </html> tag.', [ 'maybe_missing_tags' ] );
+		}
+
+		if ( ! preg_match( '/(<\/body>)/i', $html ) ) {
+			$missing_tags[] = esc_html__( '</body>', 'rocket' );
+			Logger::debug( 'Not found closing </body> tag.', [ 'maybe_missing_tags' ] );
+		}
+
+		if ( did_action( 'wp_footer' ) === 0 ) {
+			$missing_tags[] = __( 'wp_footer()', 'rocket' );
+			Logger::debug( 'Did action did not run wp_footer() function.', [ 'maybe_missing_tags' ] );
+		}
+
+		set_transient( 'rocket_missing_tags', wp_sprintf_l( '%l', $missing_tags ), HOUR_IN_SECONDS );
```

Comment:
Some concerns here:
- We should set the transient only if we detected missing tags, otherwise a page where nothing is missing will empty the transient (that may not be empty until now). Ex: visit a page where `</html>` is missing, visit a page where it isn't: no admin notice. Am I right?
- This will run on every page, so it may be a good idea to set the transient only when needed, when we have something new to record.
- What if page 1 is missing `</body>` and page 2 is missing `</html>`? Assuming they are visited in that order, only `</html>` will be shown in the admin notice.

Suggestion:
```suggestion
		if ( ! $missing_tags ) {
			return;
		}

		$transient    = get_transient( 'rocket_missing_tags' );
		$transient    = is_array( $transient ) ? $transient : [];
		$missing_tags = array_unique( array_merge( $transient, $missing_tags ) );

		if ( count( $transient ) === count( $missing_tags ) ) {
			return;
		}

		set_transient( 'rocket_missing_tags', $missing_tags, HOUR_IN_SECONDS );
```

We still have one caveat: the user doesn't know which pages are lacking which tags, but I guess it would be awful to display a notice listing hundreds of URLs ^^' (no shit Sherlock!).

Reply:
You are correct, it will be displayed only for a certain page. So if a type of page misses </body> and one misses </html> only the last visited page will display the error. 
Indeed, displaying error messages for all pages is not an option, because there will be a lot of URLs in there. 

Suggestion: Probably include the separate log files and create a log file where all pages which have missing tags are logged and display the URL to this log file in the notice?

Reply:
I first thought about storing the URLs in the transient, but it may use a lot of space in the DB.
Using a dedicated log file, why not, it’s probably the best solution so far. I think users must be able to deactivate it via filter, because of free disk space + read/write access time and resources (it's triggered on each page).

Reply:
Why do we need dedicated log file when we are already logging this into debug log? 

In 99% of the cases, most websites miss these because their footer.php doesn't have it. One page having this and the others not having it is really rare. But possible.

Reply:
I'm wondering if this improvement should be discussed some more before going further, to be sure we provide at the same time:
- The most helpful data possible
- The less obstrusively & obnoxiously possible
- Ways for the users to solve the issue themselves

Let's remember that we want this notice to be meaningful for the users, and give them the information for them to fix the issue directly. Adding stuff to the debug log doesn't help prevent a support ticket at the moment.

Reply:
You are right. I am confused at the direction this is headed to. Let me open a slack thread to see if we are all on the same page.

Reply:
Yes, this task has gone wild :)

Basically, I was thinking to provide the user the dedicated log file with all missing URLs and missing hooks in the notice. I was thinking of having the separate log file and when this issue appears, ask the client to enable debug logging and in the notice to provide them the URL to the dedicated debug file where they can see the exact missing tag from the exact URL.

Probably in this way the client will have all necessary information on how to solve by themselves the issue.